File: code/design-patterns/behavioral/iterator/go/iterator_usage.go
package main

import "fmt"

func main() {
	// Create a linked list
	head := &LinkedListNode{Value: 1}
	head.Next = &LinkedListNode{Value: 2}
	head.Next.Next = &LinkedListNode{Value: 3}
	head.Next.Next.Next = &LinkedListNode{Value: 4}
	head.Next.Next.Next.Next = &LinkedListNode{Value: 5}

	// Forward traversal
	forwardIt := NewForwardIterator(head)
	fmt.Print("Forward traversal: ")
	for value := forwardIt.Next(); value != nil; value = forwardIt.Next() {
		fmt.Print(*value, " ")
	}
	fmt.Println()

	// Backward traversal
	backwardIt := NewBackwardIterator(head)
	fmt.Print("Backward traversal: ")
	for value := backwardIt.Next(); value != nil; value = backwardIt.Next() {
		fmt.Print(*value, " ")
	}
	fmt.Println()
}

File: code/design-patterns/behavioral/iterator/python/iterator_usage.py
from iterator import ForwardIterator, BackwardIterator, LinkedListNode

head = LinkedListNode(1)
head.next = LinkedListNode(2)
head.next.next = LinkedListNode(3)
head.next.next.next = LinkedListNode(4)
head.next.next.next.next = LinkedListNode(5)

# Forward traversal
forward_iterator = ForwardIterator(head)
print("Forward traversal: ", end="")
value = forward_iterator.next()
while value is not None:
    print(value, end=" ")
    value = forward_iterator.next()
print()

# Backward traversal
backward_iterator = BackwardIterator(head)
print("Backward traversal: ", end="")
value = backward_iterator.next()
while value is not None:
    print(value, end=" ")
    value = backward_iterator.next()
print()
